src/trace_ng/log.zig

```diff
@@ -0,0 +1,672 @@
+const std = @import("std");
+const Level = @import("level.zig").Level;
+const logfmt = @import("logfmt.zig");
+const sig = @import("../sig.zig");
+// TODO Improve import.
+const Channel = @import("../sync/channel.zig").Channel;
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
+const AtomicBool = std.atomic.Value(bool);
+const RecycleFBA = sig.utils.allocators.RecycleFBA;
+
+pub const Config = struct {
+    max_level: Level = Level.debug,
+    allocator: std.mem.Allocator,
+    /// Maximum memory that logger can use.
+    max_buffer: u64,
+    exit_sig: *std.atomic.Value(bool),
+    kind: LogKind = LogKind.standard,
+};
+
+const INITIAL_LOG_CHANNEL_SIZE: usize = 1024;
+
+const LogKind = enum {
+    standard,
+    testing,
+    noop,
+};
+
+const UnscopedLogger = StandardLogger(null);
+pub fn StandardLogger(comptime scope: ?type) type {
+    const StanardErrLogger = struct {
+        const Self = @This();
+        max_level: Level,
+        exit_sig: *std.atomic.Value(bool),
+        allocator: Allocator,
+        recycle_fba: RecycleFBA,
+        max_buffer: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) Self {
+            return .{
+                .allocator = config.allocator,
+                .recycle_fba = RecycleFBA.init(config.allocator, config.max_buffer) catch @panic("could not create RecycleFBA"),
+                .max_buffer = config.max_buffer,
+                .max_level = config.max_level,
+                .exit_sig = config.exit_sig,
+                .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+                .handle = null,
+            };
+        }
+
+        fn unscoped(self: *Self) *UnscopedLogger {
+            return @ptrCast(self);
+        }
+
+        fn withScope(self: *Self, comptime new_scope: anytype) *StandardLogger(new_scope) {
+            return @ptrCast(self);
+        }
```

Comment:
Instead of this you could have pub functions called `logger` and `scopedLogger` that return the interface by value (not by pointer).
```zig
fn logger(self: *Self) Logger {
    return .{ .standard = self };
}
fn scopedLogger(self: *Self, comptime scope: []const u8) ScopedLogger(scope) {
    return .{ .standard = self };
}
```